src/targets/http.rs

```diff
@@ -76,33 +96,106 @@ impl Target {
                     "Target {}: Got update ({} entries)",
                     component.name(), update.set().len()
                 );
-                source.update(update);
+                source.update(SourceData::new(update, &mut state));
             }
         }
     }
 }
 
 
-
 //------------ Source --------------------------------------------------------
 
 /// The date source for an HTTP target.
 #[derive(Clone, Default)]
 struct Source {
     /// The current set of RTR data.
-    data: Arc<ArcSwap<Option<payload::Set>>>
+    data: Arc<ArcSwap<Option<SourceData>>>
 }
 
 impl Source {
     /// Updates the data source from the given update.
-    fn update(&self, update: payload::Update) {
-        self.data.store(Some(update.set().clone()).into())
+    fn update(&self, data: SourceData) {
+        self.data.store(Some(data).into())
     }
 
-    /// Returns the current payload set.
-    fn set(&self) -> Option<payload::Set> {
-        self.data.load().as_ref().as_ref().cloned()
+    /// Returns the current payload data.
+    fn data(&self) -> Arc<Option<SourceData>> {
+        self.data.load_full()
     }
 }
 
 
+//------------ SourceData ----------------------------------------------------
+
+/// The data held by a data source.
+///
+struct SourceData {
+    set: payload::Set,
+    etag: String,
+    created: DateTime<Utc>,
+}
+
+impl SourceData {
+    fn new(update: payload::Update, state: &mut State) -> Self {
+        let etag = format!("\"{:x}-{}\"", state.session(), state.serial());
+        state.inc();
+        Self {
+            set: update.set().clone(),
+            etag,
+            created: Utc::now(),
+        }
+    }
+
+    /// Returns whether 304 Not Modified response should be retured.
+    fn is_not_modified(&self, req: &Request<Body>) -> bool {
+        // First, check If-None-Match.
+        for value in req.headers().get_all("If-None-Match").iter() {
+            // Skip ill-formatted values. By being lazy here we may falsely
+            // return a full response, so this should be fine.
+            let value = match value.to_str() {
+                Ok(value) => value,
+                Err(_) => continue
+            };
+            let value = value.trim();
+            if value == "*" {
+                return true
+            }
+            for tag in EtagsIter::new(value) {
+                if tag.trim() == self.etag {
+                    return true
+                }
+            }
+        }
+
+        // Now, the If-Modified-Since header.
+        if let Some(value) = req.headers().get("If-Modified-Since") {
```

Comment:
According to https://www.rfc-editor.org/rfc/rfc9110#section-13.1.3:
> A recipient MUST ignore If-Modified-Since if the request contains an If-None-Match header field;

So shouldn't this code exit if it went round the first loop at least once and NOT handle the "If-Modified-Since" header in that case?